learningequality/studio
# Localize change sync operations

## Summary
### Description of the change(s) you made
* Attempts to prevent issues with background tabs sleeping timers and saving memory by making each tab responsible for the creation and persistence to the backend of edits made in the tab context
* Creates Change classes mapping to each Change type to handle proper instantiation of the change and persistence to the indexedDB changes table
* Updates all `put` operations to be `add` operations (new in Dexie v3) to make explicit create vs update changes
* Adds change creation logic into `add`, `update`, and `delete` methods
* Updates existing move, copy, publish, sync, and deploy changes to use the new Change classes
* Removes unused methods on the resource class that modify tables, to prevent future developers from using them without the appropriate change updates
* Updates the server sync logic to run in every open tab (but only to poll when the window is visible)
* Updates the debounce behaviour to remove need for the `syncActive` flag, and instead chain from previous invocations
* Updates the debounce behaviour to always return a promise that will resolve with the next sync
* Updates the forceServerSync function to try to persist all currently unsynced changes, as a last resort before someone logs out.
* Activates the service worker to be active during development (this was not ultimately needed for the implementation, and can be dropped if not wanted)

### Manual verification steps performed
1. Make edits to nodes
2. Upload new nodes
3. Open multiple tabs, make changes in each, ensure that changes from each tab sync from within the tab

___
## Reviewer guidance
### Are there any risky areas that deserve extra testing?
I think the main area for concern here is that it feels like there is more of a chance of a change not getting synced, because a page could be refreshed etc. Not sure how to make sure changes do get synced, while also maintaining the separation of tabs syncing their own changes.


## References
Hopefully fixes #3992

----

### Contributor's Checklist


PR process:


Studio-specifc:



Testing:

___

### Reviewer's Checklist
#### This section is for reviewers to fill out.
